### ISSUE TITLE
Support for versions with `+` character

### ISSUE THREAD

**Issue by serioga** [cljdoc/cljdoc#764]
I've released package with version [1.4.0+123](https://clojars.org/com.github.strojure/undertow/versions/1.4.0+123) having `+` character in it.

The cljdoc page https://cljdoc.org/d/com.github.strojure/undertow/1.4.0+123 shows 404 Not Found

> Could not find `com.github.strojure/undertow` version `1.4.0+123` in any maven repository

It looks like a bug at cljdoc side.

**Comment by lread** [comment]
Thanks for raising an issue @serioga, I'll take a peek sometime today.

**Comment by serioga** [comment]
Maybe it is not worth fixing, looks like `+` in version is problematic here and there 😞 .

**Comment by lread** [comment]
## What about Maven Central?

I was curious to learn if there are many Maven central artifacts that use the plus character in their version. This question was a bit difficult to answer. I don't know of an online Maven Central search tool that allows me to ask: give me some artifacts that have versions with + signs in them.

I settled on downloading the Maven Central indexes following the instructions I found [here](https://maven.apache.org/repository/central-index.html). For the curious, the `.gz` weighed about ~2gb, but after unpacking took up ~19gb of disk space.

The Lucene fields are described [here](https://maven.apache.org/maven-indexer-archives/maven-indexer-LATEST/indexer-core/). 

I hacked together a little Clojure to scan for versions of interest.

`deps.edn` uses an ancient version of Lucene for compatibility with downloaded index.
```Clojure
{:deps {org.apache.lucene/lucene-core {:mvn/version "4.1.0"}}}
```

`src/lucene_dump/core.clj`
```Clojure
(ns lucene-dump.core
  (:require
   [clojure.java.io :as io]
   [clojure.string :as string])
  (:import
   (org.apache.lucene.document Document)
   (org.apache.lucene.index DirectoryReader MultiFields)
   (org.apache.lucene.store FSDirectory)))

(set! *warn-on-reflection* true)

(defn scan-versions
  [index-path]
  (with-open [disk-index (FSDirectory/open (io/file index-path))
              reader (DirectoryReader/open disk-index)]
    (let [live-docs (MultiFields/getLiveDocs reader)]
      (loop [ndx 0
             of-interest []]
        (when (zero? (mod ndx 10000)) ;; this can take a while, spit out progress
          (println "ndx" ndx "of-interest" (count of-interest)))
        (if (>= ndx (.maxDoc reader))
          of-interest
          (if (and live-docs (not (.get live-docs ndx)))
            (recur (inc ndx) of-interest)
            (let [doc ^Document (.document reader ndx)
                  u (.get doc "u")
                  flds (and u (string/split u #"\|"))
                  [group-id artifact-id version] flds]
              (if (and version (string/includes? version "+"))
                (recur (inc ndx) (conj of-interest {:u u
                                                    :group-id group-id
                                                    :artifact-id artifact-id
                                                    :version version}))
                (recur (inc ndx) of-interest)))))))))

(comment
  (def of-interest (scan-versions "/home/lee/Downloads/central-lucene-index"))

  (count of-interest)
  ;; => 375559

  (first of-interest)
  ;; => {:u "io.trygvis.appsh|booter-jetty|8.1.7.v20120910+1|NA",
  ;;     :group-id "io.trygvis.appsh",
  ;;     :artifact-id "booter-jetty",
  ;;     :version "8.1.7.v20120910+1"}

  ;; so how many libs do this?
  (->> of-interest
       (group-by (juxt :group-id :artifact-id))
       count)
  ;; => 2647

  ;; let's find examples with a few characters after the some digit then + then at least few alnums
  (->> of-interest
       (filter #(re-find #"\d\+\p{Alnum}{3,}" (:version %)))
       (take-nth 200)
       (take 5))
  ;; => ({:u
  ;;      "ro.fortsoft.wicket.dashboard.widgets|wicket-dashboard-widgets-parent|0.11.0+wicket-7|NA",
  ;;      :group-id "ro.fortsoft.wicket.dashboard.widgets",
  ;;      :artifact-id "wicket-dashboard-widgets-parent",
  ;;      :version "0.11.0+wicket-7"}
  ;;     {:u "com.github.rahulsom|grooves-types|0.2.0-dev.84+3e05045|sources|jar",
  ;;      :group-id "com.github.rahulsom",
  ;;      :artifact-id "grooves-types",
  ;;      :version "0.2.0-dev.84+3e05045"}
  ;;     {:u "com.erigir|wrench-zk|2.1.0+124|javadoc|jar",
  ;;      :group-id "com.erigir",
  ;;      :artifact-id "wrench-zk",
  ;;      :version "2.1.0+124"}
  ;;     {:u "org.graylog.jest|jest-droid|2.4.12+jackson|javadoc|jar",
  ;;      :group-id "org.graylog.jest",
  ;;      :artifact-id "jest-droid",
  ;;      :version "2.4.12+jackson"}
  ;;     {:u "tech.orkestra|orkestra-lock_2.12|0.9.0+5-ad6f89b3+20180811-0338|NA",
  ;;      :group-id "tech.orkestra",
  ;;      :artifact-id "orkestra-lock_2.12",
  ;;      :version "0.9.0+5-ad6f89b3+20180811-0338"}))
```

There's about 39 million artifacts in the index, so less than 1% have a + in their versions.
Which is higher than I would have guessed.

And finally, let's bring up a pom in the browser and see how the + is treated:
https://repo1.maven.org/maven2/com/erigir/wrench-zk/2.1.0+124/wrench-zk-2.1.0+124.pom
Ya, seems fine, no need to escape, but escaping also works:
https://repo1.maven.org/maven2/com/erigir/wrench-zk/2.1.0%2B124/wrench-zk-2.1.0%2B124.pom

So, I am still currently leaning toward thinking that this is a clojars bug.
But we could easily work around it by always escaping + when consulting repos.

I'll do a wee bit more poking around. I'm curious if tools deps and lein will successfully resolve artifacts with + in their version. If not, this is probably worth a fix over at clojars.

**Comment by lread** [comment]
## TL;DR
Artifacts with versions with a `+` do not resolve if hosted on clojars, but do resolve if hosted on Maven Central. 

This means, to me, if this can and were to be fixed, should be fixed on clojars. There's not much value in cljdoc being able to build docs for your artifact, if that artifact is not resolvable using Clojure tooling.

I'll raise an issue on Clojars.

## Resolving with tools deps
Does a clojars lib resolve when using tools deps?

### [fails!] For an artifact hosted on clojars
```Clojure
❯ clojure -Sdeps '{:deps {com.github.strojure/undertow {:mvn/version "1.4.0+123"}}}' -Stree
Error building classpath. Could not find artifact com.github.strojure:undertow:jar:1.4.0+123 in central (https://repo1.maven.org/maven2/)
```
Nope!

### [works!] For an artifact hosted on Maven Central
Let's try a jar that exists on maven central:
```
❯ clojure -Sdeps '{:deps {com.erigir/wrench-zk {:mvn/version "2.1.0+124"}}}' -Stree
org.clojure/clojure 1.11.1
  . org.clojure/spec.alpha 0.3.218
  . org.clojure/core.specs.alpha 0.2.62
com.erigir/wrench-zk 2.1.0+124
  . org.zkoss.zk/zk 8.5.0
    . org.zkoss.common/zweb 8.5.0
      . org.zkoss.common/zcommon 8.5.0
        . commons-fileupload/commons-fileupload 1.3.2
          . commons-io/commons-io 2.2
        . org.zkoss.common/zel 8.5.0
          . org.javassist/javassist 3.18.2-GA
        . org.apache-extras.beanshell/bsh 2.0b6
        X org.slf4j/slf4j-api 1.7.5 :older-version
        . org.slf4j/slf4j-jdk14 1.7.5
          X org.slf4j/slf4j-api 1.7.5 :older-version
      . commons-codec/commons-codec 1.8
    . com.google.javascript/closure-compiler v20170626
      . com.google.javascript/closure-compiler-externs v20170626
      . args4j/args4j 2.33
      . com.google.errorprone/error_prone_annotations 2.0.18
      . com.google.guava/guava 20.0
      . com.google.protobuf/protobuf-java 3.0.2
      . com.google.code.gson/gson 2.7
      . com.google.code.findbugs/jsr305 3.0.1
      . com.google.jsinterop/jsinterop-annotations 1.0.0
  . org.zkoss.zk/zkbind 8.5.0
    . org.zkoss.zk/zul 8.5.0
      . org.zkoss.zk/zk 8.5.0
    . org.javassist/javassist 3.18.2-GA
  . org.slf4j/slf4j-api 1.7.25
```

## Resolving with lein

### [fails!] For an artifact hosted on clojars

`project.clj`
```
(defproject lread/testing123 "0.1.0"
  :description "Testing deps with versions with +"
  :dependencies [[com.github.strojure/undertow "1.4.0+123"]])
```

```Shell
❯ lein deps :tree
Could not find artifact com.github.strojure:undertow:jar:1.4.0+123 in central (https://repo1.maven.org/maven2/)
Could not find artifact com.github.strojure:undertow:jar:1.4.0+123 in clojars (https://repo.clojars.org/)
This could be due to a typo in :dependencies, file system permissions, or network issues.
If you are behind a proxy, try setting the 'http_proxy' environment variable.
```

### [works!] For an artifact hosted on Maven Central
`project.clj`
```Clojure
(defproject lread/testing123 "0.1.0"
  :description "Testing deps with versions with +"
  :dependencies [[com.erigir/wrench-zk "2.1.0+124"]])
```

```Clojure
❯ lein deps :tree                                                                          
 [com.erigir/wrench-zk "2.1.0+124"]
   [org.slf4j/slf4j-api "1.7.25"]
   [org.zkoss.zk/zk "8.5.0"]
     [com.google.javascript/closure-compiler "v20170626"]
       [args4j "2.33"]
       [com.google.code.findbugs/jsr305 "3.0.1"]
       [com.google.code.gson/gson "2.7"]
       [com.google.errorprone/error_prone_annotations "2.0.18"]
       [com.google.guava/guava "20.0"]
       [com.google.javascript/closure-compiler-externs "v20170626"]
       [com.google.jsinterop/jsinterop-annotations "1.0.0"]
       [com.google.protobuf/protobuf-java "3.0.2"]
     [org.zkoss.common/zweb "8.5.0"]
       [commons-codec "1.8"]
       [org.zkoss.common/zcommon "8.5.0"]
         [commons-fileupload "1.3.2"]
           [commons-io "2.2"]
         [org.apache-extras.beanshell/bsh "2.0b6"]
         [org.slf4j/slf4j-jdk14 "1.7.5" :scope "runtime"]
         [org.zkoss.common/zel "8.5.0"]
   [org.zkoss.zk/zkbind "8.5.0"]
     [org.javassist/javassist "3.18.2-GA"]
     [org.zkoss.zk/zul "8.5.0"]
 [nrepl "1.0.0" :exclusions [[org.clojure/clojure]]]
 [org.nrepl/incomplete "0.1.0" :exclusions [[org.clojure/clojure]]]
```

**Comment by lread** [comment]
All good now @serioga!
Many thanks to @tobias for working around AWS's strange S3 URL encoding: https://github.com/clojars/clojars-web/issues/862

**Comment by serioga** [comment]
hm, broken again? https://cljdoc.org/d/com.github.strojure/parsesso/1.2.2+295

> 404 - Library not found
> Could not find com.github.strojure/parsesso version 1.2.2 295 in any maven repository

**Comment by lread** [comment]
Yep, it is caused by the pedestal upgrade.

I've downgraded pedestal for now to restore `+` friendliness.

I'll leave this issue open while I figure out what's up with pedestal 0.6.0-beta-2.

**Comment by serioga** [comment]
> Since this + issue is... problematic...

let's push it to the end ;-)
